vm/devices/storage/storvsc_driver/src/lib.rs

```diff
@@ -0,0 +1,856 @@
+
+use async_channel::Receiver;
+use async_channel::RecvError;
+use async_channel::Sender;
+use futures::FutureExt;
+use futures_concurrency::future::Race;
+use guestmem::AccessError;
+use guestmem::MemoryRead;
+use guestmem::ranges::PagedRange;
+use parking_lot::Mutex;
+use std::collections::HashMap;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering;
+use task_control::AsyncRun;
+use task_control::InspectTask;
+use task_control::StopTask;
+use task_control::TaskControl;
+use thiserror::Error;
+use tracing_helpers::ErrorValueExt;
+use vmbus_async::queue;
+use vmbus_async::queue::CompletionPacket;
+use vmbus_async::queue::ExternalDataError;
+use vmbus_async::queue::IncomingPacket;
+use vmbus_async::queue::OutgoingPacket;
+use vmbus_async::queue::PacketRef;
+use vmbus_async::queue::Queue;
+use vmbus_channel::RawAsyncChannel;
+use vmbus_ring::OutgoingPacketType;
+use vmbus_ring::PAGE_SIZE;
+use vmbus_ring::RingMem;
+use vmcore::vm_task::VmTaskDriverSource;
+use zerocopy::FromBytes;
+use zerocopy::Immutable;
+use zerocopy::IntoBytes;
+use zerocopy::KnownLayout;
+
+/// Storvsc to provide a backend for SCSI devices over VMBus.
+pub struct StorvscDriver<T: Send + Sync + RingMem> {
+    storvsc: TaskControl<StorvscState, Storvsc<T>>,
+    version: storvsp_protocol::ProtocolVersion,
+    driver_source: VmTaskDriverSource,
+    new_request_sender: Option<Sender<StorvscRequest>>,
+}
+
+/// Storvsc backend for SCSI devices.
+struct Storvsc<T: Send + Sync + RingMem> {
+    inner: StorvscInner,
+    version: storvsp_protocol::ProtocolVersion,
+    queue: Queue<T>,
+    num_sub_channels: Option<u16>,
+    has_negotiated: bool,
+}
+
+struct StorvscInner {
+    next_transaction_id: AtomicU64,
+    new_request_receiver: Receiver<StorvscRequest>,
+    transactions: Mutex<HashMap<u64, PendingOperation>>,
+}
+
+struct StorvscRequest {
+    request: storvsp_protocol::ScsiRequest,
+    buf_gpa: u64,
+    byte_len: usize,
+    completion_sender: Sender<StorvscCompletion>,
+}
+
+/// Result of a Storvsc operation. If None, then operation was cancelled.
+pub struct StorvscCompletion {
+    completion: Option<storvsp_protocol::ScsiRequest>,
+}
+
+struct PendingOperation {
+    sender: Sender<StorvscCompletion>,
+}
+
+impl PendingOperation {
+    fn new(sender: Sender<StorvscCompletion>) -> Self {
+        Self { sender }
+    }
+
+    fn complete(&mut self, result: storvsp_protocol::ScsiRequest) -> Result<(), StorvscError> {
+        self.sender
+            .send_blocking(StorvscCompletion {
+                completion: Some(result),
+            })
+            .map_err(|_err| StorvscError::NotificationError)
+    }
+
+    fn cancel(&mut self) {
+        // Sending completion with an empty result indicates cancellation or other error.
+        self.sender
+            .send_blocking(StorvscCompletion { completion: None })
+            .unwrap();
+    }
+}
+
+/// Errors resulting from storvsc.
+#[derive(Debug, Error)]
+pub enum StorvscError {
```

Comment:
I suggest against making big error enums `pub`, because it prevents the compiler from pointing out variants that are unused.

The pattern I prefer is to have a top-level error type be an opaque struct, and then use an enum internally. Something like:

```rust
#[derive(Debug, Error)]
#[error(transparent)]
pub struct StorvscError(#[error(from)] StorvscInnerError);

enum StorvscInnerError { ... }
```

You may even choose to box the inner error to help the compiler avoid allocating a bunch of extra stack space for errors.

Reply:
`#[error(from)]` does not seem to be accepted by the compiler:

"expected one of: string literal, `transparent`, `fmt`"

Reply:
Oh yeah, just `#[from]`.

Reply:
I looked at other examples and found that they don't use `#[error(from)]` so I removed this and it seems to be working